typing/typedecl.ml

```diff
@@ -721,11 +848,39 @@ let check_well_founded_manifest env loc path decl =
   check_well_founded env loc path (Path.same path) visited
     (Ctype.newconstr path args)
 
+(* Given a new type declaration [type t = ...] (potentially mutually-recursive),
+   we check that accepting the declaration does not introduce ill-founded types.
+
+   Note: we check that the types at the toplevel of the declaration
+   are not reachable unguarded from themselves, that is, we check that
+   there is no cycle going through the "root" of the declaration. But
+   we *also* check that all the type sub-expressions in ..., even
+   those that occur guarded under the root of the declaration, are
+   themselves well-founded. (So we check the absence of cycles, even
+   for cycles going through inner type subexpressions but not the
+   root.
+
+   We are not actually sure that this "deep check" is necessary
+   (we don't have an example at hand where it is necessary), but we
+   are doing it anyway out of caution.
+*)
 let check_well_founded_decl env loc path decl to_check =
   let open Btype in
-  let visited = ref TypeMap.empty in
-  let checked = ref TypeSet.empty in
+  (* We iterate on all subexpressions of the declaration to check
+     "in depth" that no ill-founded type exists. *)
   let it =
+    let checked =
+      (* [checked] remembers the types that the iterator visited,
+         to avoid looping on cyclic types. *)
+      ref TypeSet.empty in
+    let visited =
+      (* [visited] remembers the inner traversals performed by
+         [check_well_founded] on each type expression reachable from
+         this declaration. This avoids unnecessary duplication of
+         [check_well_founded] work when invoked on two type
+         subexpressions that are distinct but reachable from one
+         another. *)
+      ref TypeMap.empty in
```

Comment:
Your explanation suggests that the names are interverted.
But if we change it, we should also change it in `check_well_founded`.

Reply:
I think that the current names are in fact okay, or at least that reversing them does not really improve things (the problem is that we are using two names that are almost synonyms). I rephrased the comments to avoid using "visited" when documenting "checked" and to use it when documented "visited".

Reply:
"on two type subexpressions ..." -> "on two parts of the declaration which have common subexpressions"